ENIGMAsystem/SHELL/Platforms/General/POSIX/main.cpp

```diff
@@ -23,8 +24,13 @@ void initialize_directory_globals() {
 
   // Set the program_directory
   buffer[0] = 0;
-  ssize_t count = readlink("/proc/self/exe", buffer, PATH_MAX + 1);
-  if (count !=  -1)
+  char path[PATH_MAX + 1];
+  memset(buffer, 0, sizeof(buffer));
```

Comment:
The value returned by `readlink` is the number of bytes read. There's no need to return every value. Revert this bit and just construct the string to be `count` bytes in size:

```cpp
  ssize_t count = readlink("/proc/self/exe", buffer, PATH_MAX + 1);
  if (readlink(path, buffer, PATH_MAX + 1) != -1)
    enigma_user::program_directory = add_slash(dirname(string(buffer, count)));
```

Reply:
I fixed it your royal majesty.

Reply:
@JoshDreamland to put it in your words - this code is fucking dangerous: dirname(string(buffer, count))

Reply:
dirname takes a char * argument and modifies the argument

Reply:
why do i need to construct the string to be count bytes in size

Reply:
that's pointless in this case if i know what im talking about which i may not

Reply:
Oh, `dirname` is the fucking dangerous C function. Ugh. We really need to replace that with an ENIGMA utility. In the interim, this code is fine, then:

```cpp
  ssize_t count = readlink("/proc/self/exe", buffer, PATH_MAX + 1);
  if (count != -1) {
    buffer[count] = 0;
    enigma_user::program_directory = add_slash(dirname(buffer));
  }
```

Also, don't ask me why I called `readlink` twice in my original comment. Apparently I could use to sleep more.